### PR TITLE
Fix wrong ordering of losses and add unit tests

### PR DESCRIPTION
**Pull request by florian-huber**
Originally here https://github.com/matchms/matchms-backup/pull/233

-   fix intensity order in ``add_losses``
-   add unit tests
-   add ``loss_mz_from`` and ``loss_mz_to`` parameters to ``add_losses``

Closses #20

### CODE REVIEW

File: matchms/filtering/add_losses.py [hunk 1]
```diff
@@ -17,10 +29,15 @@ def precursor_mz_is_number():
...
+        assert precursor_mz_is_number(), ("Expected 'precursor_mz' to be a scalar number.",
+                                          "Consider applying 'add_precursor_mz' filter first.")
         peaks_mz, peaks_intensities = spectrum.peaks
         losses_mz = (precursor_mz - peaks_mz)[::-1]
-        losses_intensities = peaks_intensities
-        spectrum.losses = Spikes(mz=losses_mz, intensities=losses_intensities)
+        losses_intensities = peaks_intensities[::-1]
+        # Add losses which are within given boundaries
+        keep_idx = numpy.where((losses_mz >= loss_mz_from)
+                               & (losses_mz <= loss_mz_to))[0]
```
Comment (fdiblen) [inline matchms/filtering/add_losses.py]:
```suggestion
                               & (losses_mz <= loss_mz_to))
```

File: matchms/filtering/add_losses.py [hunk 2]
```diff
@@ -17,10 +29,15 @@ def precursor_mz_is_number():
...
-        assert precursor_mz_is_number(), "Expected 'precursor_mz' to be a scalar number."
+        assert precursor_mz_is_number(), ("Expected 'precursor_mz' to be a scalar number.",
+                                          "Consider applying 'add_precursor_mz' filter first.")
         peaks_mz, peaks_intensities = spectrum.peaks
         losses_mz = (precursor_mz - peaks_mz)[::-1]
-        losses_intensities = peaks_intensities
-        spectrum.losses = Spikes(mz=losses_mz, intensities=losses_intensities)
+        losses_intensities = peaks_intensities[::-1]
+        # Add losses which are within given boundaries
+        keep_idx = numpy.where((losses_mz >= loss_mz_from)
```
Comment (fdiblen) [inline matchms/filtering/add_losses.py]:
```suggestion
        mask = numpy.where((losses_mz >= loss_mz_from)
```

File: matchms/filtering/add_losses.py [hunk 3]
```diff
@@ -17,10 +29,15 @@ def precursor_mz_is_number():
...
+                               & (losses_mz <= loss_mz_to))[0]
+        spectrum.losses = Spikes(mz=losses_mz[keep_idx],
```
Comment (fdiblen) [inline matchms/filtering/add_losses.py]:
```suggestion
        spectrum.losses = Spikes(mz=losses_mz[mask],
```

File: matchms/filtering/add_losses.py [hunk 4]
```diff
@@ -17,10 +29,15 @@ def precursor_mz_is_number():
...
+                                 intensities=losses_intensities[keep_idx])
```
Comment (fdiblen) [inline matchms/filtering/add_losses.py]:
```suggestion
                                 intensities=losses_intensities[mask])
```

File: tests/test_add_losses.py [hunk 5]
```diff
@@ -11,7 +11,10 @@ def test_add_losses():
 
     spectrum = add_losses(spectrum_in)
 
-    assert numpy.allclose(spectrum.losses.mz, numpy.array([145, 245, 295, 345], "float"))
+    aim_mz = numpy.array([145, 245, 295, 345], "float")
```
Comment (fdiblen) [inline tests/test_add_losses.py]:
```suggestion
    expected_mz = numpy.array([145, 245, 295, 345], "float")
```

File: tests/test_add_losses.py [hunk 6]
```diff
@@ -11,7 +11,10 @@ def test_add_losses():
...
+    assert numpy.allclose(spectrum.losses.mz, aim_mz), "Expected different loss m/z."
```
Comment (fdiblen) [inline tests/test_add_losses.py]:
```suggestion
    assert numpy.allclose(spectrum.losses.mz, expected_mz), "Expected different loss m/z."
```

File: tests/test_add_losses.py [hunk 7]
```diff
@@ -11,7 +11,10 @@ def test_add_losses():
...
+    aim_intensities = numpy.array([1000, 100, 200, 700], "float")
```
Comment (fdiblen) [inline tests/test_add_losses.py]:
```suggestion
    expected_intensities = numpy.array([1000, 100, 200, 700], "float")
```

File: tests/test_add_losses.py [hunk 8]
```diff
@@ -11,7 +11,10 @@ def test_add_losses():
...
+    assert numpy.allclose(spectrum.losses.intensities, aim_intensities), "Expected different intensities."
```
Comment (fdiblen) [inline tests/test_add_losses.py]:
```suggestion
    assert numpy.allclose(spectrum.losses.intensities, expected_intensities), "Expected different intensities."
```

File: tests/test_add_losses.py [hunk 9]
```diff
@@ -48,3 +51,29 @@ def test_add_losses_with_input_none():
...
+
+
+def test_add_losses_with_peakmz_larger_precursormz():
+    spectrum_in = Spectrum(mz=numpy.array([100, 150, 200, 450], dtype="float"),
+                           intensities=numpy.array([700, 200, 100, 1000], dtype="float"),
+                           metadata={"precursor_mz": 445.0})
+
+    spectrum = add_losses(spectrum_in)
+
+    aim_mz = numpy.array([245, 295, 345], "float")
```
Comment (fdiblen) [inline tests/test_add_losses.py]: rename

**Review by fdiblen — CHANGES_REQUESTED** [review]
Thanks

File: tests/test_add_losses.py (hunk 9 above)
Comment (florian-huber) [reply tests/test_add_losses.py]: I renamed it accordingly. Thanks!

File: matchms/filtering/add_losses.py [hunk 10]
```diff
@@ -17,10 +29,15 @@ def precursor_mz_is_number():
 
     precursor_mz = spectrum.get("precursor_mz")
     if precursor_mz is not None:
```
Comment (fdiblen) [inline matchms/filtering/add_losses.py]:
```suggestion
    if precursor_mz:
```

File: tests/test_add_losses.py [hunk 11]
```diff
@@ -11,7 +11,10 @@ def test_add_losses():
 
     spectrum = add_losses(spectrum_in)
 
-    assert numpy.allclose(spectrum.losses.mz, numpy.array([145, 245, 295, 345], "float"))
+    expected_mz = numpy.array([145, 245, 295, 345], "float")
+    assert numpy.allclose(spectrum.losses.mz, expected_mz), "Expected different loss m/z."
+    expected_intensities = numpy.array([1000, 100, 200, 700], "float")
```
Comment (fdiblen) [inline tests/test_add_losses.py:16]: please move this to just after line 14

File: tests/test_add_losses.py [hunk 12]
```diff
@@ -48,3 +51,29 @@ def test_add_losses_with_input_none():
...
+def test_add_losses_with_peakmz_larger_precursormz():
+    spectrum_in = Spectrum(mz=numpy.array([100, 150, 200, 450], dtype="float"),
+                           intensities=numpy.array([700, 200, 100, 1000], dtype="float"),
+                           metadata={"precursor_mz": 445.0})
+
+    spectrum = add_losses(spectrum_in)
+
+    expected_mz = numpy.array([245, 295, 345], "float")
+    assert numpy.allclose(spectrum.losses.mz, expected_mz), "Expected different loss m/z."
+    expected_intensities = numpy.array([100, 200, 700], "float")
```
Comment (fdiblen) [inline tests/test_add_losses.py:69]: the same here: after line 63

File: tests/test_add_losses.py [hunk 13]
```diff
@@ -48,3 +51,29 @@ def test_add_losses_with_input_none():
     spectrum_in = None
     spectrum = add_losses(spectrum_in)
     assert spectrum is None
+
+
+def test_add_losses_with_peakmz_larger_precursormz():
```
Comment (fdiblen) [inline tests/test_add_losses.py:60]: Can you please add a single sentence to tests explaining what you test here?

**Review by fdiblen — CHANGES_REQUESTED** [review]
Thanks @florian-huber! I suggested a few more very minor changes.

File: tests/test_add_losses.py (hunk 11 above)
Comment (florian-huber) [reply tests/test_add_losses.py:16]: Sorry, I am not sure what you mean here.

File: tests/test_add_losses.py (hunk 13 above)
Comment (florian-huber) [reply tests/test_add_losses.py:60]: done

File: tests/test_add_losses.py (hunk 11 above)
Comment (florian-huber) [reply tests/test_add_losses.py:16]: changed the order. hope that's what you meant?

File: tests/test_add_losses.py (hunk 12 above)
Comment (florian-huber) [reply tests/test_add_losses.py:69]: changed the order. hope that's what you meant?

File: tests/test_add_losses.py (hunk 11 above)
Comment (fdiblen) [reply tests/test_add_losses.py:16]: yes, thanks

**Review by fdiblen — APPROVED** [review]
good to go. Thanks!

Edit: Of course after CI checks :))

**Comment by florian-huber** [comment]
Thanks for reviewing Faruk.

**Pull request merged**